Sources/SwiftParser/Expressions.swift

```diff
@@ -319,17 +319,42 @@ extension Parser {
         return nil
       }
     case (.arrow, _)?, (.throwsKeyword, _)?:
-      let asyncKeyword = self.consumeIfContextualKeyword("async")
+      var asyncKeyword = self.consumeIfContextualKeyword("async")
+
+      var throwsKeyword = self.consume(if: .throwsKeyword)
 
-      let throwsKeyword = self.consume(if: .throwsKeyword)
       let (unexpectedBeforeArrow, arrow) = self.expect(.arrow)
 
+      // Recover if effect modifiers are specified after '->' by eating them into
+      // unexpectedAfterArrow and inserting the corresponding effects modifier as
+      // missing into the ArrowExprSyntax. This reflect the semantics the user
+      // originally intended.
+      var effectModifiersAfterArrow: [RawTokenSyntax] = []
```

Comment:
This only handles `throws async` and not eg.`throws throws async`. I've put up a very partial PR for function signatures (https://github.com/apple/swift-syntax/pull/889). Any thoughts on something like that instead? There's quite a few places we'll want to change this 😅 (if we decide this is in fact something the parser should handle).